lrauv_ignition_plugins/test/test_ahrs.cc

```diff
@@ -0,0 +1,527 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/*
+ * Development of this module has been funded by the Monterey Bay Aquarium
+ * Research Institute (MBARI) and the David and Lucile Packard Foundation
+ */
+
+#include <gtest/gtest.h>
+
+#include <ignition/common/Console.hh>
+#include <ignition/common/Filesystem.hh>
+#include <ignition/gazebo/components/AngularVelocityCmd.hh>
+#include <ignition/gazebo/components/LinearVelocityCmd.hh>
+#include <ignition/gazebo/components/Physics.hh>
+#include <ignition/gazebo/components/Pose.hh>
+#include <ignition/gazebo/components/PoseCmd.hh>
+#include <ignition/gazebo/components/Recreate.hh>
+#include <ignition/gazebo/components/Static.hh>
+#include <ignition/gazebo/Model.hh>
+#include <ignition/gazebo/TestFixture.hh>
+#include <ignition/gazebo/Types.hh>
+#include <ignition/gazebo/Util.hh>
+#include <ignition/gazebo/World.hh>
+#include <ignition/math/Pose3.hh>
+#include <ignition/math/Quaternion.hh>
+#include <ignition/math/Vector3.hh>
+#include <ignition/msgs/imu.pb.h>
+#include <ignition/msgs/magnetometer.pb.h>
+#include <ignition/msgs/Utility.hh>
+#include <ignition/transport/Node.hh>
+
+#include <chrono>
+#include <functional>
+#include <memory>
+#include <string>
+#include <optional>
+#include <utility>
+
+#include "TestConstants.hh"
+
+using namespace std::literals::chrono_literals;
+
+class AhrsTest : public ::testing::Test
+{
+  protected: void SetUp() override {
+    ignition::common::Console::SetVerbosity(4);
+
+    this->fixture = std::make_unique<ignition::gazebo::TestFixture>(
+        ignition::common::joinPaths(std::string(PROJECT_SOURCE_PATH),
+                                    "worlds", "buoyant_tethys.sdf"));
+
+    this->node.Subscribe(
+        "/model/tethys/ahrs/imu", &AhrsTest::OnImuMessage, this);
+
+    this->node.Subscribe(
+        "/model/tethys/ahrs/magnetometer",
+        &AhrsTest::OnMagnetometerMessage, this);
+
+    this->fixture->OnConfigure(
+      [&](const ignition::gazebo::Entity &,
+          const std::shared_ptr<const sdf::Element> &,
+          ignition::gazebo::EntityComponentManager &_ecm,
+          ignition::gazebo::EventManager &)
+      {
+        ignition::gazebo::World world(
+            ignition::gazebo::worldEntity(_ecm));
+        ignition::gazebo::Entity tethysEntity =
+            world.ModelByName(_ecm, "tethys");
+        using ignition::gazebo::components::Static;
+        _ecm.CreateComponent(tethysEntity, Static(true));
```

Comment:
@braanan @arjo129 I forced the model to be static to avoid motion whenever I put the AUV in unusual positions. Unfortunately, it also means we are not testing AHRS output when moving. It'd be tricky nonetheless. Hydrodynamics modelling gets in the way and it's hard to tell what the angular velocity and linear acceleration is going to be.

Reply:
I think we agreed that as long as the AHRS is reporting the correct coordinate frame we should hope for the best with the gyro.

Reply:
That's right. Thing is, I cannot test for gyro frame conventions if it's output is always zero. I guess we _are_ testing that the body frame is correct for the accelerometer so it should be good for the gyro too.